python/cvmfs/repository.py

```diff
@@ -198,38 +205,223 @@ def _pop_catalog(self):
 
 
 
-class Repository:
-    """ Abstract Wrapper around a CVMFS Repository representation """
+class Cache:
+    def __init__(self, cache_dir):
+        if not os.path.exists(cache_dir):
+            cache_dir = tempfile.mkdtemp(dir='/tmp', prefix='cache.')
+        self._cache_dir = cache_dir
+        self._create_cache_structure()
+        self._cleanup_metadata()
+
+    def _cleanup_metadata(self):
+        for f in glob.glob(self._cache_dir + '/' + '.cvmfs*'):
+            os.remove(f)
+
+
+    def _create_dir(self, path):
+        cache_full_path = os.path.join(self._cache_dir, path)
+        if not os.path.exists(cache_full_path):
+            os.mkdir(cache_full_path, 0755)
+
+
+    def _create_cache_structure(self):
+        self._create_dir('data')
+        for i in range(0x00, 0xff + 1):
+            new_folder = '{0:#0{1}x}'.format(i, 4)[2:]
+            self._create_dir(os.path.join('data', new_folder))
+
+
+    def cache_path(self):
+        return str(self._cache_dir)
+
+
+    def add(self, file_name):
+        full_path = os.path.join(self._cache_dir, file_name)
```

Comment:
In the long run that could become a concurrency problem if two 'things' try to add the same file. The C++ client does that transactionally to get around it, like so:

1. Open a temporary file in **data/txn** (for TransaXioN)
2. Let the user write the file
3. Call `commit()` to move the file in place **data/aa/....**

That way two process can `add()` the same file at the same time and since `rename()` is atomic in POSIX, their `commit()`s will just overwrite each other which is harmless.